# Profitable App Profiles for the App Store and Google Play Markets

<br />

<br />


![header](header.jpg)


## Introduction


Working for a company that **only build free apps to download and install** available on Google Play and in the App Store, our main source of revenue consists of in-app ads. 

This means that the number of users of our apps determines our revenue for any given app therefore the more users who see and engage with the ads, the better. 

<br />

**Introduction to the Project**


In today's mobile development landscape, it is crucial to understand which types of applications attract more users in both Android and Apple ecosystems. With the goal of providing valuable information to developers, this project focuses on data analysis to identify common characteristics among the applications that receive a higher number of downloads and engagement.

1. **What Does the Project Entail?**

    This data analysis combines information from various sources, such as Google Play Store and iTunes App Store, along with usage and performance metrics. Using statistical techniques and machine learning algorithms, patterns and trends are identified to indicate which types of applications perform better in each platform.

2. **Project Objective**

    The primary objective of this project is to provide developers with precise information about the characteristics that resonate best with users on Android and iOS. By understanding better what types of applications perform well in terms of downloads, engagement, and retention, developers can make more informed decisions for the design and optimization of their own apps.


<br />


Our aim is collect and analyze data from each of the online stores to understand what type of apps it would be interesting to develop to attract more users.

## <a id='0'>Index</a>

 
### <a href='#1'>1. Exploring data </a>
 
 
- <a href='#11'>1.1 Function to explore datasets</a>


-  <a href='#12'>1.2 Data dictionary:</a>


-  <a href='#13'>1.3 Knowing the size in both data sets.</a>

### <a href='#2'>2. Data cleaning </a>


-  <a href='#21'>2.1. Detect inaccurate data... remove it.</a>


-  <a href='#22'>2.2 Detect duplicate data, and remove the duplicates.</a>


-  <a href='#23'>2.3 What differentiates duplicate applications?</a>


-  <a href='#24'>2.4 Removing Non-English Apps</a>


-  <a href='#25'>2.5 Filter non-English applications from both datasets. </a>


-  <a href='#26'>2.6 Exploring datasets and see how many rows we have left for each dataset.</a>



### <a href='#3'>3. Isolating Free Apps </a>


-  <a href='#31'>3.1 Our **ultimate goal** </a>


-  <a href='#32'>3.2 Profiles of apps successful in both markets	Table</a>



### <a href='#4'>4. Most Popular Apps by Genre on Apple Store </a>

### <a href='#5'>5. Most Popular Apps by Genre on Apple Store </a>

### <a href='#6'>5. Summary. </a>

* * *

The datasets are in two documents of type csv's i.e. **Comma Separated Values file** for read the content, it is necessary to load a function called `reader` of the python `csv` module.

In [1]:
!pwd

/home/ion/Documentos/albertjimrod/data-projects/01_Python/00-Profitable App Profiles for the App Store and Google Play Markets


In [2]:
from csv import reader

* * *
<a href='#0'> back to index</a>

## <a id='1'>1. Exploring datasets</a>:

We are going to explore the two sets of data, but first we are going to kwon what type of [character encoding](https://en.wikipedia.org/wiki/Character_encoding) both datasets have.

A command called [file](https://www.man7.org/linux/man-pages/man1/file.1.html), will help us to kwon the type of file regardless of its extension and avoid a error called `UnicodeDecodeError`.


```
FILE(1)                                         BSD General Commands Manual                                         FILE(1)

NAME
     file — determine file type

SYNOPSIS
     file [-bcdEhiklLNnprsSvzZ0] [--apple] [--exclude-quiet] [--extension] [--mime-encoding] [--mime-type] [-e testname]
          [-F separator] [-f namefile] [-m magicfiles] [-P name=value] file ...
     file -C [-m magicfiles]
     file [--help]

DESCRIPTION
     This manual page documents version 5.41 of the file command.

     file tests each argument in an attempt to classify it.  There are three sets of tests, performed in this order:
     filesystem tests, magic tests, and language tests.  The first test that succeeds causes the file type to be printed.

     The type printed will usually contain one of the words text (the file contains only printing characters and a few com‐
     mon control characters and is probably safe to read on an ASCII terminal), executable (the file contains the result of
     compiling a program in a form understandable to some UNIX kernel or another), or data meaning anything else (data is
     usually “binary” or non-printable).  Exceptions are well-known file formats (core files, tar archives) that are known
     to contain binary data.  When modifying magic files or the program itself, make sure to preserve these keywords.
     Users depend on knowing that all the readable files in a directory 

In [3]:
! file -i csv/AppleStore.csv

csv/AppleStore.csv: text/csv; charset=utf-8


In [4]:
! file -i csv/googleplaystore.csv

csv/googleplaystore.csv: text/csv; charset=utf-8


Our files have the same type of charset https://en.wikipedia.org/wiki/UTF-8.


Quick idea:

- **charset:** is the set of characters you can use.
- **encoding:** is the way these characters are stored into memory.

[source](https://stackoverflow.com/questions/2281646/whats-the-difference-between-encoding-and-charset)

In [5]:
AppleStore = open('csv/AppleStore.csv', encoding='utf8')
AppleStore = reader(AppleStore)
AppleStore = list(AppleStore)
header_apple = [x for x in AppleStore[0]]
header_apple #Columns on AppleStore

['id',
 'track_name',
 'size_bytes',
 'currency',
 'price',
 'rating_count_tot',
 'rating_count_ver',
 'user_rating',
 'user_rating_ver',
 'ver',
 'cont_rating',
 'prime_genre',
 'sup_devices.num',
 'ipadSc_urls.num',
 'lang.num',
 'vpp_lic']

In [6]:
dataset_apple = AppleStore[1:] #whole dataset apple, no headers

In [7]:
GooglePlay = open('csv/googleplaystore.csv', encoding='utf8')
GooglePlay = reader(GooglePlay)
GooglePlay = list(GooglePlay)
header_google = [x for x in GooglePlay[0]]
header_google 

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [8]:
dataset_google = GooglePlay[1:] #whole dataset google, no header

* * *
<a href='#0'> back to index</a>

### <a id='11'>1.1 Function to explore datasets</a>

One dataset contains data from approximately 10,000 **Android apps from Google Play**; Data was collected in August 2018. One dataset contains data from approximately 7,000 **iOS apps from the App Store**; Data was collected in July 2017. Let's open and explore these datasets with the 'explore_data()' function Parameters: 

<br>

- 'dataset': A list of lists that represents the dataset.
- 'start': An integer that indicates the initial index of the segment to be displayed.
- 'end': An integer that indicates the final (exclusive) index of the segment to be displayed.
- 'rows_and_columns' (optional): A Boolean with a default value of False. If True, it will print the number of rows and columns.

To make easier to explore, we created a function named `explore_data()` you can repeatedly use to print rows in a readable way.

In [9]:
def explore_data(dataset, start, end, rows_and_columns=False):
        dataset_slice = dataset[start:end]    
        for row in dataset_slice:
            print(row)
            print('\n') # adds a new (empty) line after each row
        if rows_and_columns:
            print('Number of rows:', len(dataset))
            print('Number of columns:', len(dataset[0]))

In [10]:
%%html
<style>
table {float:left}
</style>

<a href='#0'> back to index</a>
* * *
### <a id='12'> 1.2 Data dictionary:</a>

We have the name of the columns and the data that form the set of both data sets:


#### AppleStore.csv

 - `header_apple`

 - `dataset_apple` 

||Name | Description|
|:--|:---|:--|
|1|"id" : |App ID|
|2|"track_name": |App Name|
|3|"size_bytes": |Size (in Bytes)|
|4|"currency": |Currency Type|
|5|"price": |Price amount|
|6|"ratingcounttot": |User Rating counts (for all version)|
|7|"ratingcountver": |User Rating counts (for current version)|
|8|"user_rating" : |Average User Rating value (for all version)|
|9|"userratingver": |Average User Rating value (for current version)|
|10|"ver" : |Latest version code|
|11|"cont_rating": |Content Rating|
|12|"prime_genre": |Primary Genre|
|13|"sup_devices.num": |Number of supporting devices|
|14|"ipadSc_urls.num": |Number of screenshots showed for display|
|15|"lang.num": |Number of supported languages|
|16|"vpp_lic": |Vpp Device Based Licensing Enabled|


#### Googleplaystore.csv

 - `header_google`

 - `dataset_google`


||Name | Description|
|:--|:---|:--|
|1 | App: |Application name|
|2 | Category: |Category the app belongs to|
|3 | Rating: |Overall user rating of the app (as when scraped)|
|4 | Reviews: | Number of user reviews for the app (as when scraped)|
|5 | Size: | Size of the app (as when scraped)|
|6 | Installs: | Number of user downloads/installs for the app (as when scraped)|
|7 | Type: | Paid or Free|
|8 | Price: | Price of the app (as when scraped)|
|9 | Content: | Rating Age group the app is targeted at - Children / Mature 21+ / Adult|
|10 |  Genres: | An app can belong to multiple genres (apart from its main category). For eg, a musical family game will belong to Music, Game, Family genres.|

<a href='#0'> back to index</a>
* * *
### <a id='13'>1.3 Knowing the size in both data sets</a>.

In [11]:
explore_data(dataset_apple, 2, 4, True)

['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7197
Number of columns: 16


In [12]:
explore_data(dataset_google, 2, 4, True)

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10841
Number of columns: 13


### Resume

|`dataset_apple`|`dataset_google`|
|:----|:----|
|Number of rows: 7197| Number of rows: 10841|
|Number of columns: 16|Number of columns: 13|

<a href='#0'> back to index</a>
* * *
## <a id='2'>2. Data cleaning</a>

This means that we need to:

- 1. **Detect inaccurate data, and correct or remove it**.

- 2. **Detect duplicate data, and remove the duplicates**.


- 3. **Remove non-English apps like 爱奇艺PPS -《欢乐颂2》电视剧热播**.

- 4. **Remove apps that aren't free##**.


The Google Play dataset has a dedicated [discussion section](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion?sort=votes), and we can see that [one of the discussions](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion/66015) describes an error for a certain row.

<a href='#0'> back to index</a>
* * *
### <a id='21'>2.1. Detect inaccurate data... remove it.</a>

If the number of columns in a row varies then some data is missing.

This function `field_number` has a pair of parameters `dataset`, `header` in which we enter the dataset and the header.

The passage of each row compare the number of fields with the number of fields in header if they do not coincide then in that line **some type of data is missing**.

In [14]:
def field_number(dataset, header):
    index_number = 0
    for row in range(len(dataset)):
        if len(dataset[row]) != len(header):
            print(dataset[row])
            print("\n")
            print("error in index number: ",index_number)
        index_number +=1

In [15]:
field_number(dataset_apple, header_apple)

In [16]:
dataset_google[0]

['Photo Editor & Candy Camera & Grid & ScrapBook',
 'ART_AND_DESIGN',
 '4.1',
 '159',
 '19M',
 '10,000+',
 'Free',
 '0',
 'Everyone',
 'Art & Design',
 'January 7, 2018',
 '1.0.0',
 '4.0.3 and up']

In [17]:
field_number(dataset_google, header_google)

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


error in index number:  10472


In [18]:
len(dataset_google[10472]) #discussion row

12

In [19]:
dataset_google[10472]

['Life Made WI-Fi Touchscreen Photo Frame',
 '1.9',
 '19',
 '3.0M',
 '1,000+',
 'Free',
 '0',
 'Everyone',
 '',
 'February 11, 2018',
 '1.0.19',
 '4.0 and up']

We have passed the two datasets through the function and there is a row in the `dataset_google` with a different length from the number of columns therefore we are going to remove it from our dataset.

In [20]:
del dataset_google[10472]

Checking if row has been removed properlly.

In [21]:
dataset_google[10472]

['osmino Wi-Fi: free WiFi',
 'TOOLS',
 '4.2',
 '134203',
 '4.1M',
 '10,000,000+',
 'Free',
 '0',
 'Everyone',
 'Tools',
 'August 7, 2018',
 '6.06.14',
 '4.4 and up']

#### yes ...It has been removed

<a href='#0'> back to index</a>
* * *
### <a id='22'>2.2 Detect duplicate data, and remove the duplicates.</a>

`for` is running through each row extracting the name of the application.

- Checking if it already exists or not.

- if not exist then the name is inserted in `unique_apps` else is inserted into `duplicate_apps`.

In [22]:
def detect_duplicated(dataset):
    unique_apps = []
    duplicate_apps = []
    
    for row in dataset:
        name = row[0]
        if name in unique_apps:
            duplicate_apps.append(name)
        else:
            unique_apps.append(name)
            
    if len(duplicate_apps) == 0:
        print("In this dataset there are no duplicate apps")
    else:
        times=len(duplicate_apps)
        print("There are {reapeated} repeated apps".format(reapeated=times,dataset=dataset))
        print("\n")
        print('Examples of duplicate apps:',"\n"+"\n" , duplicate_apps[:5])

In [23]:
detect_duplicated(dataset_apple)

In this dataset there are no duplicate apps


In [24]:
detect_duplicated(dataset_google)

There are 1181 repeated apps


Examples of duplicate apps: 

 ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


Just only 5 samples of the repeated applications in **dataset_google**

* * * 
<a href='#0'> back to index</a>
### <a id='23'>2.3 What differentiates duplicate applications?</a>

If we take as an example a repeated application, in this case `Instagram`, we see that of all the fields that make up the row **there is only one that is different**.

In [26]:
for app in dataset_google[1:]:
    if app[0] == 'Quick PDF Scanner + OCR FREE':
        print(app)

['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80804', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']


In [27]:
for app in dataset_google[1:]:
    if app[0] == 'MailChimp - Email, Marketing Automation':
        print(app)

['MailChimp - Email, Marketing Automation', 'BUSINESS', '4.1', '5448', '12M', '500,000+', 'Free', '0', 'Everyone', 'Business', 'July 25, 2018', '4.9.1', '5.0 and up']
['MailChimp - Email, Marketing Automation', 'BUSINESS', '4.1', '5448', '12M', '500,000+', 'Free', '0', 'Everyone', 'Business', 'July 25, 2018', '4.9.1', '5.0 and up']


### Contextual Explanation

When we analyze a dataset of applications, it's common to find some applications listed multiple times. To address this, we need to determine the best criteria to remove these duplicates.

#### Initial Observation:
Once we have verified that there are many applications (some repeated more than once), we need to decide how to handle the duplicates. Specifically, we want to retain only those versions of an application that have the most recent and relevant data.

#### Criterion for Removing Duplicates:
The criterion chosen is: "The higher the number of reviews, the more recent the data should be!" 
Instead of randomly removing duplicate entries, we will keep only those applications that are repeated and have the highest number of reviews. The versions with fewer reviews will be eliminated in favor of those with more.

#### Function to Create a Dictionary:
To implement this criterion, we use a function that processes a dataset (usually a list of lists) and creates a dictionary where:
- **Key:** The name of the app.
- **Value:** The maximum number of reviews associated with that app.

This approach ensures that for each application, we keep only the version with the highest review count. 

#### Creating the Dictionary `reviews_max`:
We create a dictionary called `reviews_max`, where:
- The key is the name of the application.
- The value is the number of reviews for that application (extracted from the data).

To populate this dictionary, we follow these steps:
1. **Extracting Data**: We extract the `name` and `n_reviews` (the number of reviews) for each application.
2. **Updating or Adding to the Dictionary**:
   - If a name already exists in the dictionary and the value associated with that key is less than the current `n_reviews`, we update the value.
   - If a name does not exist in the dictionary, we add it along with its corresponding `n_reviews`.

This ensures that our final dictionary contains only the most reviewed versions of each application.

### Example:
Suppose we have the following data for three applications:

- App A: 10 reviews
- App B: 20 reviews (repeated twice)
- App C: 5 reviews

After processing, our `reviews_max` dictionary would look like this:
- Key: "App A", Value: 10 reviews
- Key: "App B", Value: 20 reviews
- Key: "App C", Value: 5 reviews

Thus, for each application, we keep the version with the highest number of reviews.

### Summary:
By using this approach, we ensure that our dataset contains only the most relevant and reviewed versions of each application, making it more accurate and useful.

In [ ]:
reviews_max = {} #key=name, value=review

for row in dataset_google:
    name = row[0]
    n_reviews = float(row[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews: #if name in reviews_max & value in reviews_max[name]
        reviews_max[name] = n_reviews                         #is < than n_reviews the update
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print('length of the dictionary:',len(reviews_max))

This is the new length of our dataset `length of the dictionary: 9659` having deleted repeated applications.

The original length of the dataset minus the number of repeated applications results in this number.

In [ ]:
reviews_max

In [ ]:
print('expected length', len(dataset_google) - 1181) #1181 repeated apps

- Filter the maximum number of apps reviews with the dictionary and that the name is not added.

In [ ]:
android_clean = []
already_added = []

for row in dataset_google: #original dataset
    name = row[0]
    n_reviews = float(row[3])
    
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(row) #row list added to list
        already_added.append(name) #new name added list

`android_clean` is the our **new dataset**

In [ ]:
explore_data(android_clean, 1, 2, True)

<a href='#0'> back to index</a>
* * *
### <a id='24'>2.4 Removing Non-English Apps</a>

It is no necessary to do the same with App Store data because there are no duplicates.

We use English for the applications we develop in our company, and we would like to analyze only applications aimed at an English-speaking audience. 


However, if we explore the data enough, we will find that **both datasets** have applications with names that suggest they are not aimed at an English-speaking audience.

In [ ]:
print(AppleStore[813][1])
print(AppleStore[6731][1])
print('\n')
print(android_clean[4412][0])
print(android_clean[7940][0])

**One way to do this is to remove all applications whose name contains a symbol that is not commonly used in English text**. 

English text typically includes letters of the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;) and other símbols (+, *, /).

Each character we use in a string has a corresponding number associated with it. 

For example, the corresponding number of the character"a" es 97, el del carácter "A" es 65, y el del carácter "爱" es 29.233. 

 We can get the corresponding number of each character using the built-in function **ord().**

In [ ]:
print(ord('a'))
print(ord('A'))
print(ord('爱'))
print(ord('5'))
print(ord('+'))

The numbers corresponding to the characters that we usually use in an English text are all in the range of 0 to 127, according to the ASCII system (American Standard Code for Information Interchange). 

Based on this range of numbers, we can construct a function that detects whether a character belongs to the common English character set or not. 

- **If the number is equal to or less than 127, the character belongs to the set of common English characters. 

- **If an app's name contains a character greater than 127, it probably means that the app has a non-English name.** 

This is the basis of operation. 

Read each of the characters and compare their ASCII value with the one we have determined as limit (127).

In [ ]:
def english_speak(string):
    for character in string:
        valor_character = ord(character)
        if valor_character > 127:
            return False #not English 
        else:
            return True # is English

In [ ]:
english_speak('Instagram')

In [ ]:
english_speak('爱奇艺PPS -《欢乐颂2》电视剧热播')

In [ ]:
english_speak('Docs To Go™ Free Office Suite')#Is English!

In [ ]:
english_speak('Instachat 😜') #Is English!

In [ ]:
print(ord('™'))
print(ord('😜'))

If we are going to use the function we have created, we will lose useful data, we saw that the function could not correctly identify certain names of English applications such as **'Docs To Go™ Free Office Suite'** and **'Instachat 😜'** since many English applications will be incorrectly labeled as non-English. 


To **minimize the impact of data loss**,it is necessary to have a basic criterion that helps in the screening so **we will only delete an application if its name is longer than three characters with the corresponding numbers outside the ASCII range**. 

This means that all English apps with up to three emoji or other special characters will still be labeled as English. 

In [ ]:
def english_speak(text):
    non_validchar = [] #aux list
    for character in text:
        valor_character = ord(character)
        if valor_character > 127:
            non_validchar.append(valor_character)
    if len(non_validchar) >= 3:
        return False
    else:
        return True

In [ ]:
english_speak('Docs To Go™ Free Office Suite')

In [ ]:
english_speak('爱奇艺PPS -《欢乐颂2》电视剧热播')

In [ ]:
english_speak('Instachat 😜 😜')

In [ ]:
english_speak('Instachat 😜 😜 😜')

At this point it is important to remember what the names of the datasets are, thus avoiding failures.

|Dataset names|lenght|
|:----|:----|
|`dataset_apple`|7197|
|`android_clean`|9659, 8757|

<a href='#0'> back to index</a>
* * *
### <a id='25'>2.5 Filter non-English applications from both datasets.</a> 

If an app's name is identified as English, the we add the entire row to a separate list.

In [ ]:
# AppleStore

list_apple_clean = []

for row in dataset_apple:
    name = row[1]
    if english_speak(name):
        list_apple_clean.append(row)
    else:
        pass

In [ ]:
## android_clean

list_android_clean = []

for row in android_clean:
    nombre = row[0]
    if english_speak(nombre):
        list_android_clean.append(row)
    else:
        pass

<a href='#0'> back to index</a>
* * *
### <a id='26'>2.6 Exploring datasets and see how many rows we have left for each dataset.</a>

In each cleaning process we have put different names to our dataset variables so it is important to take them into account

In [ ]:
explore_data(list_apple_clean, 2, 4, True)

In [ ]:
explore_data(list_android_clean, 2, 4, True)

**Now we are working with this:**


|Dataset names|lenght|
|:----|:----|
|`list_apple_clean`|6155|
|`list_android_clean`|9597|

<a href='#0'> back to index</a>
* * *
## <a id='3'>3. Isolating Free Apps</a>

Our datasets contain **free and non-free applications** we need to isolate only the free applications for our analysis.

Checking the length of each dataset to see how many apps have left.

In [ ]:
header_apple

In [ ]:
free_apple_clean = []

for row in list_apple_clean:
    prix = row[4] #'price'
 
    if prix == '0.0' or prix == '0': #not numbers just characters! 
        free_apple_clean.append(row)
    else:
        pass
    
len(free_apple_clean)

In [ ]:
header_google

In [ ]:
free_android_clean = [] 

for row in list_android_clean:
    prix = row[7] #'Price'
    if prix == '0.0' or prix == '0': #not numbers just characters! 
        free_android_clean.append(row)
    else:
        pass
    
len(free_android_clean)

**Updating names and lengths**


|Dataset names|lenght|
|:----|:----|
|`free_apple_clean`|3203|
|`free_android_clean`|8848|

<a href='#0'> back to index</a>
* * *

### <a id='31'>3.1 Our ultimate goal</a>

Our **ultimate goal** is to add the app on both Google Play and the App Store, so we need to **find profiles of apps that are successful in both markets**. 

Let's start the analysis by having an idea of what are the most common genres for each market. To do this, we will have to build frequency tables for some columns of our datasets.

The columns that can give us the information we need are:

`free_apple_clean[11]` ---> `prime_genre`

`free_android_clean[]` ---> `Genres`

In [ ]:
header_apple

In [ ]:
header_google

In [ ]:
def freq_table(dataset, column):
    table = {}
    total = 0
    
    for row in dataset:
        total +=1 # counter give us length.
        value = row[column]
        if value in table:
            table[value] +=1
        else:
            table[value] = 1
        
    table_percentages = {}
    
    for key in table: #table[column value]<--as key] value <--as +1 values added 
        percentage = (table[key] / total ) * 100
        table_percentages[key]=percentage
            
    return table_percentages # {column:times repeated}

In [ ]:
print(freq_table(free_apple_clean, 11))

In [ ]:
print(freq_table(free_android_clean, 9))


- The value of the dictionary we place in `key_val_as_tuple`, tuple along with the key of this


In [ ]:
def display_table(dataset, column):
    table = freq_table(dataset, column) #return diccionary
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)   #times repeated & key same variable
        table_display.append(key_val_as_tuple) #append tuple into list

    table_sorted = sorted(table_display, reverse = True) #sorting list
    for entry in table_sorted: #print every cell on list
        print(entry[1], ':', round(entry[0],2),'%')

In [ ]:
output_apple = display_table(free_apple_clean,11) #apple prime_genere column

In [ ]:
output_gogle = display_table(free_android_clean,9)

<a href='#0'> back to index</a>
* * *
### <a id='32'>3.2 Profiles of apps successful in both markets</a>

|Dataset name|Genres|percentage|
|:---|:---|:---|
|`free_apple_clean`|||
||Games : |58 %|
||**Entertainment :** |7,84 %|
||Photo & Video : |5 %|
||Education : |4 %|
||Social Networking : |3 %|
|`free_android_clean`|||
||Tools : 8 %
||**Entertainment :**| 6.08 %|
||Education : |5 %|
||Business : |5 %|
||Productivity : |4 %|

### We would like to get an idea about the type of apps with more users.

One way to find out which genres are the most popular (have the most users) is to calculate the **average number of installs for each app genre**.

<br>

- For the **Google Play** dataset, we can find this information in the **installs** column, 


- For the **App Store dataset** we will take the total number of user ratings as a proxy, which we can find in the `'rating_count'` column app.

In [ ]:
header_apple[5]

In [ ]:
for idx in range(5):
    print(free_apple_clean[idx][5])

These are the number of installations per application

***

In [ ]:
header_google[5]

In [ ]:
for idx in range(5):
    print(free_android_clean[idx][5])

...and this is an indicative number and therefore it is very important to note that the data contained in these columns is of a different nature.

Let's start with the apple set.

<a href='#0'> back to index</a>
* * *

### <a id='5'>5. Most Popular Apps by Genre on Apple Store ¶</a>

In [ ]:
print(free_apple_clean[1][:])

In [ ]:
genres_ios = freq_table(free_apple_clean, -5)
list_order = {}

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in free_apple_clean:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
            
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)
    
    list_order[genre]=avg_n_ratings

### - sorting Genres on App Store 

In [ ]:
marklist = sorted(list_order.items(), key=lambda x:x[1],reverse = True)
sortdict = dict(marklist)
sortdict

## <a id='4'>4. Most Popular Apps by Genre on Android</a>

In [ ]:
for app in free_apple_clean:
    if app[-5] == 'Navigation':    # number one
        print(app[1], ':', app[5]) # print name and number of ratings

The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

**Our aim** is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. 

**The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold**. 


We could get a **better picture by removing these extremely popular apps** for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:

### What applications are there in the second most popular genre?

In [ ]:
for app in free_apple_clean:
    if app[-5] == 'Reference': # number two
        print(app[1], ':', app[5])

This niche seems to show some potential. 

One thing we could do is take a popular book and turn it into an app where we could add different features besides the raw version of the book: 

- This might include daily quotes from the book.

- an audio version of the book.

- quizzes about the book, etc. 

- we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

<br>

This idea seems to fit well with the fact that the **App Store is dominated by for-fun apps**. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting:

- Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

- Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

- Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

- Therefore our premise is to avoid the big guys.



<a href='#0'> back to index</a>
* * *
## <a id='5'>5. Most Popular Apps by Genre on Google Play</a>

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

## Analyze the Google Play market

In [ ]:
display_table(free_android_clean, 5) # the Installs columns

One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [ ]:
categories_android = freq_table(free_android_clean, 1)
list_order={}

for category in categories_android:
    total = 0
    len_category = 0
    for app in free_android_clean:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)
    
    list_order[category]=avg_n_installs

In [ ]:
marklist = sorted(list_order.items(), key=lambda x:x[1],reverse = True)
sortdict = dict(marklist)
sortdict

On average, communication apps have the most installs: 38590581.08741259. 

**This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts)**, and a few others with over 100 and 500 million installs:

In [ ]:
for app in free_android_clean:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+' or 
                                      app[5] == '500,000,000+' or 
                                      app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times:

In [ ]:
under_100_m = []

for app in free_android_clean:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:

In [ ]:
for app in free_android_clean:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [ ]:
for app in free_android_clean:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [ ]:
for app in free_android_clean:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

<a href='#0'> back to index</a>
* * *
## <a id='6'>6. Summary
In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.